File: release-watcher/handler.py
import os
import json
import requests
import psycopg


def read_secret(name):
    try:
        with open(f"/var/openfaas/secrets/{name}") as f:
            lines = [l for l in f.read().splitlines() if l.strip() and not l.strip().startswith("#")]
            return lines[0].strip() if lines else ""
    except FileNotFoundError:
        return ""


def get_db_conn(dsn):
    return psycopg.connect(dsn)


def ensure_table(conn):
    with conn.cursor() as cur:
        cur.execute("""
            CREATE TABLE IF NOT EXISTS release_state (
                repo       TEXT PRIMARY KEY,
                last_tag   TEXT NOT NULL,
                updated_at TIMESTAMPTZ DEFAULT now()
            )
        """)
    conn.commit()


def get_last_tag(conn, repo):
    with conn.cursor() as cur:
        cur.execute("SELECT last_tag FROM release_state WHERE repo = %s", (repo,))
        row = cur.fetchone()
        return row[0] if row else None


def upsert_last_tag(conn, repo, tag):
    with conn.cursor() as cur:
        cur.execute("""
            INSERT INTO release_state (repo, last_tag, updated_at)
            VALUES (%s, %s, now())
            ON CONFLICT (repo) DO UPDATE
                SET last_tag   = EXCLUDED.last_tag,
                    updated_at = now()
        """, (repo, tag))
    conn.commit()


def fetch_latest_release(repo):
    url = f"https://api.github.com/repos/{repo}/releases/latest"
    headers = {"Accept": "application/vnd.github+json"}
    resp = requests.get(url, headers=headers, timeout=10)
    resp.raise_for_status()
    return resp.json()


def post_discord(webhook_url, repo, release):
    tag = release.get("tag_name", "unknown")
    html_url = release.get("html_url", "")
    body = release.get("body") or ""
    description = body[:300] + ("…" if len(body) > 300 else "")

    payload = {
        "embeds": [
            {
                "title": f"New release: {repo} {tag}",
                "description": description or "_No release notes provided._",
                "url": html_url,
                "color": 0x3CB371,
                "footer": {"text": "release-watcher · OpenFaaS"},
            }
        ]
    }
    resp = requests.post(webhook_url, json=payload, timeout=10)
    resp.raise_for_status()


def handle(event, context):
    repo = os.getenv("WATCH_REPO", "openfaas/faas")
    discord_webhook_url = read_secret("discord-webhook-url")
    pg_conn_string = read_secret("pg-conn")

    release = fetch_latest_release(repo)
    except requests.HTTPError as e:
        return {
            "statusCode": 502,
            "body": json.dumps({"error": f"GitHub API error: {e}"}),
            "headers": {"Content-Type": "application/json"},
        }

    latest_tag = release.get("tag_name")
    if not latest_tag:
        return {
            "statusCode": 502,
            "body": json.dumps({"error": "GitHub response missing tag_name"}),
            "headers": {"Content-Type": "application/json"},
        }

    # Compare against stored state
    try:
        conn = get_db_conn(pg_conn_str)
        ensure_table(conn)
        last_tag = get_last_tag(conn, repo)
    except Exception as e:
        return {
            "statusCode": 500,
            "body": json.dumps({"error": f"Postgres error: {e}"}),
            "headers": {"Content-Type": "application/json"},
        }

    notified = False
    if latest_tag != last_tag:
        try:
            post_discord(discord_webhook_url, repo, release)
        except requests.HTTPError as e:
            conn.close()
            return {
                "statusCode": 502,
                "body": json.dumps({"error": f"Discord webhook error: {e}"}),
                "headers": {"Content-Type": "application/json"},
            }
        upsert_last_tag(conn, repo, latest_tag)
        notified = True

    conn.close()

    return {
        "statusCode": 200,
        "body": json.dumps({
            "repo": repo,
            "latest_tag": latest_tag,
            "previous_tag": last_tag,
            "notified": notified,
        }),
        "headers": {"Content-Type": "application/json"},
    }

File: release-watcher/handler.go
package function

import (
	"bytes"
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"strings"
	"time"

	_ "github.com/lib/pq"
)

const userAgent = "release-monitor"

var (
	repo              string
	discordWebhookURL string
	db                *sql.DB
	githubClient      *http.Client
)

func init() {
	repo = getEnv("watch_repo", "openfaas/faas")
	log.Printf("Watching repository: %s", repo)
	discordWebhookURL = readSecret("discord-webhook-url")
	postgresPassword := readSecret("postgres-passwd")

	// Build Postgres connection string from environment variables
	pgHost := getEnv("postgres_host", "postgresql")
	pgPort := getEnv("postgres_port", "5432")
	pgUser := getEnv("postgres_user", "postgres")
	pgDB := getEnv("postgres_db", "postgres")
	pgConnStr := fmt.Sprintf("host=%s port=%s user=%s password=%s dbname=%s sslmode=disable",
		pgHost, pgPort, pgUser, postgresPassword, pgDB)

	// Initialize connection pool
	var err error
	db, err = sql.Open("postgres", pgConnStr)
	if err != nil {
		log.Fatalf("Error connecting to database during init: %v", err)
	}

	// Set connection pool limits to prevent connection exhaustion attacks
	db.SetMaxOpenConns(2)            // Max 2 concurrent connections
	db.SetMaxIdleConns(1)            // Keep 1 idle connection ready
	db.SetConnMaxLifetime(time.Hour) // Recycle connections after 1 hour

	// Ensure table exists on startup
	if err := ensureTable(db); err != nil {
		log.Fatalf("Error creating table during init: %v", err)
	}

	// Initialize HTTP client for GitHub (no redirect following)
	githubClient = &http.Client{
		Timeout: 10 * time.Second,
		CheckRedirect: func(req *http.Request, via []*http.Request) error {
			return http.ErrUseLastResponse
		},
	}
}

type DiscordWebhook struct {
	Content string `json:"content"`
}

func Handle(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()

	// Fetch latest releaseURL url from GitHub
	releaseURL, err := fetchLatestReleaseURL(ctx, repo)
	if err != nil {
		log.Printf("Error fetching release: %v", err)
		http.Error(w, fmt.Sprintf("Error fetching release: %v", err), http.StatusInternalServerError)
		return
	}

	tagName := releaseURL[strings.LastIndex(releaseURL, "/")+1:]

	// Get last seen tag from database
	lastTag, err := getLastTag(db, repo)
	if err != nil {
		log.Printf("Error getting last tag: %v", err)
		http.Error(w, fmt.Sprintf("Database query error: %v", err), http.StatusInternalServerError)
		return
	}

	if lastTag == tagName {
		msg := fmt.Sprintf("No new release for %s (current: %s)", repo, tagName)
		log.Println(msg)
		w.Write([]byte(msg))
		return
	}

	// Post to Discord
	if err := postToDiscord(ctx, discordWebhookURL, releaseURL); err != nil {
		log.Printf("Error posting to Discord: %v", err)
		http.Error(w, fmt.Sprintf("Discord webhook error: %v", err), http.StatusInternalServerError)
		return
	}

	// Update last seen tag
	if err := updateLastTag(db, repo, tagName); err != nil {
		log.Printf("Error updating last tag: %v", err)
		http.Error(w, fmt.Sprintf("Database update error: %v", err), http.StatusInternalServerError)
		return
	}

	msg := fmt.Sprintf("Posted new release %s for %s to Discord", tagName, repo)
	log.Println(msg)
	w.Write([]byte(msg))
}

func getEnv(key, defaultValue string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return defaultValue
}

func readSecret(name string) string {
	data, err := os.ReadFile(fmt.Sprintf("/var/openfaas/secrets/%s", name))
	if err != nil {
		log.Printf("Warning: could not read secret %s: %v", name, err)
		return ""
	}
	return string(bytes.TrimSpace(data))
}

func fetchLatestReleaseURL(ctx context.Context, repo string) (string, error) {
	url := fmt.Sprintf("https://github.com/%s/releases/latest", repo)
	req, err := http.NewRequestWithContext(ctx, http.MethodHead, url, nil)
	if err != nil {
		return "", err
	}
	req.Header.Set("User-Agent", userAgent)

	res, err := githubClient.Do(req)
	if err != nil {
		return "", err
	}
	defer res.Body.Close()

	if res.StatusCode != http.StatusFound {
		io.Copy(io.Discard, res.Body)
		return "", fmt.Errorf("GitHub returned uunexpected status: %d", res.StatusCode)
	}

	releaseURL := res.Header.Get("Location")
	if len(releaseURL) == 0 {
		return "", fmt.Errorf("unable to determine latest release")
	}

	return releaseURL, nil
}

func ensureTable(db *sql.DB) error {
	query := `
		CREATE TABLE IF NOT EXISTS release_state (
			repo TEXT PRIMARY KEY,
			last_tag TEXT NOT NULL,
			updated_at TIMESTAMPTZ DEFAULT NOW()
		)
	`
	_, err := db.Exec(query)
	return err
}

func getLastTag(db *sql.DB, repo string) (string, error) {
	var lastTag string
	err := db.QueryRow("SELECT last_tag FROM release_state WHERE repo = $1", repo).Scan(&lastTag)
	if err == sql.ErrNoRows {
		return "", nil
	}
	return lastTag, err
}

func updateLastTag(db *sql.DB, repo, tag string) error {
	query := `
		INSERT INTO release_state (repo, last_tag, updated_at)
		VALUES ($1, $2, NOW())
		ON CONFLICT (repo) DO UPDATE SET last_tag = $2, updated_at = NOW()
	`
	_, err := db.Exec(query, repo, tag)
	return err
}

func postToDiscord(ctx context.Context, webhookURL string, release string) error {
	webhook := DiscordWebhook{
		Content: release,
	}

	payload, err := json.Marshal(webhook)
	if err != nil {
		return err
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, webhookURL, bytes.NewBuffer(payload))
	if err != nil {
		return err
	}
	req.Header.Set("User-Agent", userAgent)
	req.Header.Set("Content-Type", "application/json")

	res, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	defer res.Body.Close()

	if res.StatusCode >= 400 {
		body, _ := io.ReadAll(res.Body)
		return fmt.Errorf("Discord webhook returned %d: %s", res.StatusCode, string(body))
	}

	return nil
}
